File: summer-school/MPI/jacobi/jacobi-to-parallelise.f90
!
! Jacobi iteration with Blocking Sends and Receives 
! Modified from Archer Training Material (EPCC)
!
! http://www.archer.ac.uk/training/course-material/2016/09/160929_AdvMPI_EPCC/index.php
! 
program jacobi
  use mpi
  implicit none     

  ! Boundary values
  real(kind=8), parameter :: TOP = 1.0 
  real(kind=8), parameter :: BOTTOM = 10.0 
  real(kind=8), parameter :: LEFT = 1.0 
  real(kind=8), parameter :: RIGHT = 1.0

  ! The convergence to terminate at
  real(kind=8), parameter :: CONVERGENCE_ACCURACY = 1e-4
  ! The maximum number of iterations
  integer, parameter :: MAX_ITERATIONS = 500000
  ! How often to report the norm
  integer, parameter :: REPORT_NORM_PERIOD = 1000

  real(kind=8), dimension(:,:), allocatable :: grid, grid_new, temp
  real(kind=8) :: bnorm, rnorm, norm, tmpnorm
  integer :: i, j, k, ierr, size, myrank, local_nx, nx, ny, requests(4)
  integer status(MPI_STATUS_SIZE)
  character(len=32) :: arg
  double precision :: start_time

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, myrank, ierr)
  start_time=MPI_Wtime()

  if (myrank ==0 .and. command_argument_count() /= 2) then
     print *, &
          "You must provide two command line arguments, the global size in X and the global size in Y"
     stop
  end if

  call get_command_argument(1, arg)
  read(arg,*) nx
  call get_command_argument(2, arg)
  read(arg,*) ny

  if (myrank==0) then
     print *, "Global size in X=", nx, "Global size in Y=", ny
  end if

  local_nx=nx/size
  if (local_nx * size .lt. nx) then
     if (myrank .lt. nx - local_nx * size) local_nx=local_nx+1
  end if

  allocate(grid(0:ny+1, 0:local_nx+1), grid_new(0:ny+1, 0:local_nx+1), temp(0:ny+1, 0:local_nx+1))

  bnorm=0.0
  tmpnorm=0.0
  rnorm=0.0

  ! intialise values
  if (myrank ==0) then
     grid(:,0)=LEFT
  else
     grid(:,0)=0.0d0
  endif
  if (myrank ==size-1) then
     grid(:,local_nx+1)=RIGHT
  else
     grid(:,local_nx+1)=0.d0
  endif

  ! top and lower borders
  grid(0,:)=TOP
  grid(ny+1,:)=BOTTOM

  ! updateable grid points
  grid(1:ny,1:local_nx)=0.d0
  grid_new=grid


  do j=1, local_nx
     do i=1, ny
        tmpnorm=tmpnorm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1)) ** 2)
     end do
  end do
  call mpi_allreduce(tmpnorm, bnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)
  bnorm=sqrt(bnorm)


  do k=0, MAX_ITERATIONS

     ! Copy boundaries into halo regions
     ! 2 conditions needed here for the left (rank 0) and right boundaries (rank
     ! size-1)

     ! send first data column into right halo region of myank-1, recv last data column from myrank-1 into
     ! left halo (not needed for rank 0)

     ! send last data column into right halo regioni of myrank+1, recv first data column from
     ! myrank+1 into  left halo (not needed for rank=size-1)

     tmpnorm=0.0
     do j=1, local_nx
        do i=1, ny
           tmpnorm=tmpnorm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1)) ** 2)
        end do
     end do
     call mpi_allreduce(tmpnorm, rnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)
     norm=sqrt(rnorm)/bnorm
     if (norm .lt. CONVERGENCE_ACCURACY) exit

     do j=1, local_nx
        do i=1, ny
           grid_new(i,j)=0.25 * (grid(i-1,j) + grid(i+1,j) + grid(i,j-1) + grid(i,j+1))
        end do
     end do
     
     ! set current grid to new grid
     grid=grid_new

     if (mod(k, REPORT_NORM_PERIOD)==0 .and. myrank==0) print *, "Iteration=",k," Relative Norm=",norm
  end do
  if (myrank==0) print *, "Terminated on ",k," iterations, Relative Norm=", norm, " runtime=",MPI_Wtime()-start_time," sec"
  deallocate(grid, grid_new)
  call mpi_finalize(ierr)
end program jacobi

File: summer-school/MPI/jacobi/jacobi-to-parallelise.c
//  
//   Inspired by Advanced MPI (Archer Training Material,EPCC)
//  
//   http://www.archer.ac.uk/training/course-material/2016/09/160929_AdvMPI_EPCC/index.php
//   
//
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "mpi.h"

// Boundary values 
#define TOP 1.0
#define BOTTOM 10.0
#define LEFT 1.0
#define RIGHT 1.0
//
// The maximum number of iterations
#define MAX_ITERATIONS 5000000
// The convergence to terminate at
#define CONVERGENCE_ACCURACY 1e-4
// How often to report the norm
#define REPORT_NORM_PERIOD 1000

int nx, ny, ny2;

void initialise(double**, double**, int, int, int);
double* allocate_matrix_as_array(int nrows, int ncols);
double** allocate_matrix(int nrows, int ncols, double* arr_A);

int main(int argc, char * argv[]) {	
	int size, myrank;

	MPI_Init(&argc, &argv);
	MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
	MPI_Comm_size(MPI_COMM_WORLD, &size);

	if (argc != 3) {
		if (myrank==0) fprintf(stderr, "You must provide the size in X and size in Y as arguments to this code\n");
		return -1;
	}
	nx=atoi(argv[1]);
	ny=atoi(argv[2]);
        ny2=ny+2;

	if (myrank==0) printf("Solving to accuracy of %.0e, global system size is x=%d y=%d\n", CONVERGENCE_ACCURACY, nx, ny);
	int local_nx=nx/size;
	if (local_nx * size < nx) {
		if (myrank < nx - local_nx * size) local_nx++;
	}

	double * grid1d =  allocate_matrix_as_array(local_nx+2,ny+2);
	double * grid_new1d = allocate_matrix_as_array(local_nx+2,ny+2);

        double **grid = allocate_matrix(local_nx+2,ny+2,grid1d);
        double **grid_new = allocate_matrix(local_nx+2,ny+2,grid_new1d);
	double start_time;

        MPI_Win win;
        MPI_Aint disp;

	initialise(grid, grid_new, local_nx, myrank, size);

	double rnorm=0.0, bnorm=0.0, norm, tmpnorm=0.0;
	MPI_Request requests[]={MPI_REQUEST_NULL, MPI_REQUEST_NULL, MPI_REQUEST_NULL, MPI_REQUEST_NULL};

        // Initial nor factor
	int i,j,k;
	for (i=1;i<=local_nx;i++) {
		for (j=1;j<ny+1;j++) {		
			tmpnorm=tmpnorm+pow(grid[i][j]*4-grid[i][j-1]-grid[i][j+1]-grid[i-1][j]-grid[i+1][j],2);
		}
	}
	MPI_Allreduce(&tmpnorm, &bnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
	bnorm=sqrt(bnorm);

        printf("bnorm=%lf\n",bnorm);
	start_time=MPI_Wtime();
	for (k=0;k<MAX_ITERATIONS;k++) {
 
	
               // Here you need to arange the halo exchange neighbouring domains
               // Consider 2 cases: domains with rank > 0  and rank < size-1	

     // Copy boundaries into halo regions
     

    //  send first data row into right halo region of myrank-1, recv last data row from myrank-1 into
    //  left halo (not needed for rank 0)

    //  send last data row into right halo regioni of myrank+1, recv first data row from
    //  myrank+1 into  left halo (not needed for rank <size-1)

		tmpnorm=0.0;
		for (i=1;i<=local_nx;i++) {
			for (j=1;j<ny+1;j++) {
				tmpnorm=tmpnorm+pow(grid[i][j]*4-grid[i][j-1]-grid[i][j+1]-grid[i-1][j]-grid[i+1][j],2);
			}
		}
		MPI_Allreduce(&tmpnorm, &rnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
		norm=sqrt(rnorm)/bnorm;
		if (norm < CONVERGENCE_ACCURACY) break;		
		for (i=1;i<=local_nx;i++) {
			for (j=1;j<ny+1;j++) {
				grid_new[i][j]=0.25*(grid[i][j-1]+grid[i][j+1]+grid[i-1][j]+grid[i+1][j]);
			}
		}
		memcpy(grid1d, grid_new1d, sizeof(double) * (local_nx + 2) * ny2);

		if (k % REPORT_NORM_PERIOD == 0 && myrank==0) printf("Iteration= %d Relative Norm=%e\n", k, norm);
	}
	if (myrank==0) printf("\nTerminated on %d iterations, Relative Norm=%e, Total time=%e seconds\n", k, norm,
			MPI_Wtime() - start_time);
	free(grid1d);
	free(grid_new1d);
	free(grid);
	free(grid_new);
	MPI_Finalize();
	return 0;
}

/**
 * Initialises the arrays, such that grid contains the boundary conditions at the start and end points and all other
 * points are zero. grid_new is set to equal grid
 */
void initialise(double ** grid, double ** grid_new, int local_nx, int myrank, int size) {
	int i, j;

        for (j=0;j<=local_nx+1;j++) {
                grid_new[j][0]=grid[j][0]=LEFT;
        }
        for (j=0;j<=local_nx+1;j++) {
                grid_new[j][ny2-1]=grid[j][ny2-1]=RIGHT;
        }

	for (j=0;j<=ny+1;j++) {
		grid_new[0][j]=grid[0][j]=myrank==0 ? TOP: 0;
	}
	for (j=0;j<=ny+1;j++) {
		grid_new[local_nx+1][j]=grid[local_nx+1][j]= myrank==size-1 ? BOTTOM: 0;		
	}
	for (i=1;i<=local_nx;i++) {
		for (j=1;j<ny+1;j++) {
			grid_new[i][j]=grid[i][j]=0;
		}
	}	
}


/* Allocate a double matrix with one malloc */
double* allocate_matrix_as_array(int nrows, int ncols) {
  double *arr_A;
  /* Allocate enough raw space */
  arr_A = (double*) malloc(nrows*ncols*sizeof(double));
  return arr_A;
}

/* Easy way to access a 1d dynamic array */
double** allocate_matrix(int nrows, int ncols, double* arr_A) {
  double **A;
  int ii;
  /* Prepare pointers for each matrix row */
  A = (double**) malloc(sizeof(double)*nrows);
  /* Initialize the pointers */
  for (ii=0; ii<nrows; ++ii) {
    A[ii] = &(arr_A[ii*ncols]);
  }
  return A;
}

